Make some integer conversions in JFR explicit

thallium opened the pull request:
To avoid build failure on Windows with JFR enabled.

thallium commented:
[Internal Jenkins build](https://hyc-runtimes-jenkins.swg-devops.com/job/Build_JDK21_x86-64_windows_Personal/308/) not sure why it failed on building DDR but it finished compiling C/C++ code. I'm building again with latest changes.

tajila commented on runtime/vm/JFRChunkWriter.cpp:
```diff
@@ -107,7 +107,7 @@ VM_JFRChunkWriter::writeJFRMetadata()
 
 	/* add metadata size */
-	_bufferWriter->writeLEB128PaddedU32(dataStart, _bufferWriter->getCursor() - dataStart);
+	_bufferWriter->writeLEB128PaddedU32(dataStart, U_32(_bufferWriter->getCursor() - dataStart));
```
please use (U_32)(_bufferWriter->getCursor() - dataStart), Im not sure all the compilers will support the notation you have

thallium replied:
Done, and since https://github.com/eclipse-openj9/openj9/pull/20040 is merged, also fixed conversion introduced in that PR.

tajila commented:
jenkins test sanity xlinux jdk17

tajila commented:
jenkins test sanity zlinux64 jdk17

tajila commented:
jenkins test sanity alinux64 jdk17

tajila commented:
There is an error here

```
14:42:17  [ 55%] Building CXX object runtime/gc_modron_standard/CMakeFiles/j9modronstandard_full.dir/StandardAccessBarrier.cpp.o
14:42:17  In file included from /home/jenkins/workspace/Build_JDK17_x86-64_linux_Personal/openj9/runtime/vm/JFRWriter.hpp:34,
14:42:17                   from /home/jenkins/workspace/Build_JDK17_x86-64_linux_Personal/openj9/runtime/vm/jfr.cpp:31:
14:42:17  /home/jenkins/workspace/Build_JDK17_x86-64_linux_Personal/openj9/runtime/vm/JFRChunkWriter.hpp: In member function 'void VM_JFRChunkWriter::writeIntermediateJFRChunkToFile()':
14:42:17  /home/jenkins/workspace/Build_JDK17_x86-64_linux_Personal/openj9/runtime/vm/JFRChunkWriter.hpp:223:41: error: format '%llX' expects argument of type 'long long unsigned int', but argument 4 has type 'UDATA' {aka 'long unsigned int'} [-Werror=format=]
14:42:17    223 |                 sprintf(fileName, "%s%llX.jfr", intermediateChunkFileName, (U_64)_vm->jfrState.jfrChunkCount);
14:42:17        |                                      ~~~^                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
14:42:17        |                                         |                                  |
14:42:17        |                                         long long unsigned int             UDATA {aka long unsigned int}
14:42:17        |                                      %lX
14:42:19  cc1plus: all warnings being treated as errors
```

thallium commented:
So this is a little bit tricky as `UDATA` is `long long unsigned int` on windows. I guess I'll have to cast to this to `U_32`.

tajila commented:
Keep it as `%lx` and add a `long int` cast

tajila commented:
jenkins test sanity alinux64 jdk17

tajila commented:
jenkins test sanity xlinux jdk17

thallium commented:
Looks like failed on `error: RPC failed; curl 55 TCP connection reset by peer`.

tajila commented:
Actually, I think if you use `%zx` you dont need a cast

tajila commented:
please rebase this PR

keithc-ca commented on runtime/vm/JFRChunkWriter.hpp:
```diff
@@ -232,7 +232,7 @@ class VM_JFRChunkWriter {
 		const size_t fileNameLen = sizeof(intermediateChunkFileName) + 16 + sizeof(".jfr");
 		char fileName[fileNameLen];
-		snprintf(fileName, fileNameLen, "%s%lX.jfr", intermediateChunkFileName, _vm->jfrState.jfrChunkCount);
+		snprintf(fileName, fileNameLen, "%s%zX.jfr", intermediateChunkFileName, _vm->jfrState.jfrChunkCount);
```
This should use `OMR_PRIXPTR`.

thallium commented:
I will need some time to fix implicit conversions introduced by recent PRs.

thallium commented:
Rebased and fixed implicit conversions introduced by recent PRs

tajila commented:
@thallium Please rebase these changes

tajila commented:
@keithc-ca Please take another look at this

keithc-ca commented on runtime/vm/JFRChunkWriter.cpp:
```diff
@@ -109,7 +109,7 @@ VM_JFRChunkWriter::writeJFRMetadata()
 
 	/* add metadata size */
-	_bufferWriter->writeLEB128PaddedU32(dataStart, _bufferWriter->getCursor() - dataStart);
+	_bufferWriter->writeLEB128PaddedU32(dataStart, (U_32)(_bufferWriter->getCursor() - dataStart));
```
Simply casting to `U_32` seems wrong to me.

This pattern occurs often enough that I suggest `VM_BufferWriter` should have a method to write the size:
```c
		_bufferWriter->writeSize(dataStart);
```
That method can use `this->getCursor()` to compute the size, validate that it fits in 28 bits and store it at `dataStart`.

BTW, the `U32` suffix might lead one to believe that sizes can be up to 32 bits wide, but that's not true: methods named with the `U32` suffix only support values up to 28 bits wide. Similar comments apply for the `U64` and `U72` suffixes. I think we should consider renaming those methods.

keithc-ca replied:
To expand on the comment above: `VM_BufferWriter` should treat writing values that don't fit in the related format as a form of "overflow".

tajila replied:
I dont think we need both U_64 and U_72 so we can drop usages of U_64 and just use U_72 there.

tajila replied:
In the `writeLEB128PaddedU32` check that the input has the 4 MSB set to zero, if not `_overflow = true;`

keithc-ca replied:
> I think `padded[Size]` makes it clear that the actual data is less than the memory consumed.

On the contrary, "padded" suggests to me that the output is _padded_ as necessary; the "U32" part says to me that the argument is 32 bits wide, which is not true.

tajila commented:
@thallium Please rebase this PR

thallium commented:
I've updated my PR to include the helper functions. Also updated the commit message to better reflect the changes. @tajila @keithc-ca could you take another look?

keithc-ca commented on runtime/vm/jfr.cpp:
```diff
@@ -965,16 +965,16 @@ jfrCPULoad(J9VMThread *currentThread)
 				int64_t timeDelta = currentTime - jfrState->prevProcTimestamp;
-				jfrEvent->jvmUser = OMR_MIN((currentProcCPUTimes._userTime - jfrState->prevProcCPUTimes._userTime) / ((double)numberOfCpus * timeDelta), 1.0);
-				jfrEvent->jvmSystem = OMR_MIN((currentProcCPUTimes._systemTime - jfrState->prevProcCPUTimes._systemTime) / ((double)numberOfCpus * timeDelta), 1.0);
+				jfrEvent->jvmUser = OMR_MIN((currentProcCPUTimes._userTime - jfrState->prevProcCPUTimes._userTime) / ((float)numberOfCpus * timeDelta), 1.0f);
+				jfrEvent->jvmSystem = OMR_MIN((currentProcCPUTimes._systemTime - jfrState->prevProcCPUTimes._systemTime) / ((float)numberOfCpus * timeDelta), 1.0f);
```
`OMR_MIN` evaluates one of its arguments twice; suggest using a local to avoid that.

tajila commented on runtime/vm/JFRChunkWriter.hpp:
```diff
@@ -193,6 +194,38 @@ class VM_JFRChunkWriter {
 	}
 
+	U_8 *reserveCheckpointSize() {
+		if (J9JAVAVM_COMPRESS_OBJECT_REFERENCES(_vm)) {
```
Format allows 32 or 72but sizes to be used interchangeably

keithc-ca replied:
Sure, but where is the information that will allow a reader to know how many bits are needed to read the "checkpoint size"?

tajila replied:
Its a variable integer, so its read one byte at a time regardless of whether its 4bytes or 9bytes.

keithc-ca replied:
Right, the high bit of each byte signals where there are more byte (or not).

tajila commented on runtime/vm/JFRChunkWriter.hpp:
```diff
@@ -193,6 +194,38 @@ class VM_JFRChunkWriter {
+		if (J9JAVAVM_COMPRESS_OBJECT_REFERENCES(_vm)) {
+			_bufferWriter->writeLEB128PaddedU32(checkpointEventStart, (U_32)(_bufferWriter->getCursor() - checkpointEventStart));
+		} else {
+			_bufferWriter->writeLEB128PaddedU72(checkpointEventStart, (U_32)(_bufferWriter->getCursor() - checkpointEventStart));
```
this one should have a U32 cast

tajila commented on runtime/vm/JFRChunkWriter.hpp:
```diff
@@ -193,6 +194,38 @@ class VM_JFRChunkWriter {
+	}
+
+	static void writeEventSize(VM_BufferWriter *bufferWriter, U_8 *eventStart) {
+		bufferWriter->writeLEB128PaddedU72(eventStart, (U_32)(bufferWriter->getCursor() - eventStart));
```
and here

tajila commented on runtime/vm/JFRChunkWriter.hpp:
```diff
@@ -193,6 +194,38 @@ class VM_JFRChunkWriter {
+	}
+
+	void writeCheckpointSize(U_8 *checkpointEventStart) {
+		if (J9JAVAVM_COMPRESS_OBJECT_REFERENCES(_vm)) {
```
you can make a local for this

thallium replied:
In `VM_JFRChunkWriter`?

tajila replied:
no in JFRChunkWritier

thallium commented on runtime/vm/BufferWriter.hpp:
```diff
@@ -316,6 +316,10 @@ class VM_BufferWriter {
 	void
 	writeLEB128PaddedU32(U_32 val)
```
would you mind explaining what is a blind cast?

keithc-ca commented on runtime/vm/BufferWriter.hpp:
```diff
@@ -316,6 +316,10 @@ class VM_BufferWriter {
 	void
 	writeLEB128PaddedU32(U_32 val)
 	{
+		if ((val >> 28) != 0) {
```
The constant belongs on the left, suggest `return` is not needed:
```c
		if (0 != (val >> 28)) {
			_overflow = true;
		} else if (checkBounds(4)) {
```

An equivalent overflow test should be added to `writeLEB128PaddedU64()`.

The bounds check in `writeLEB128()` should be `9`, not `sizeof(U_64)`. Suggest the arguments should just be integer literals as they express the number of bytes of space required.

thallium replied:
The bounds check in `writeLEB128()` is already 9. If you mean `writeLEB128PaddedU64()`, I think it should be 8?

keithc-ca replied:
Yes, I see `writeLEB128()` checks for 9 bytes - I don't remember how I saw something different, and yes, the check in `writeLEB128PaddedU64()` should be for 8 bytes.

keithc-ca commented on runtime/vm/JFRChunkWriter.cpp:
```diff
@@ -109,10 +109,10 @@ VM_JFRChunkWriter::writeJFRMetadata()
 
 	if (_debug) {
-		j9tty_printf(PORTLIB, "Metadata size = %d, fromLEB128 =%u\n", (int)(_bufferWriter->getCursor() - dataStart), VM_BufferWriter::convertFromLEB128ToU32(dataStart));
+		j9tty_printf(PORTLIB, "Metadata size = %zu, fromLEB128 =%u\n", _bufferWriter->getCursor() - dataStart, VM_BufferWriter::convertFromLEB128ToU32(dataStart));
```
This line (and several other calls to `j9tty_printf()`) are too long; please fold.

keithc-ca replied:
There's a missing space after `=` in the format here.

Also the cast to `int` and the use of `%d` should be fixed on line 101.

keithc-ca commented on runtime/vm/JFRChunkWriter.hpp:
```diff
@@ -193,6 +194,38 @@ class VM_JFRChunkWriter {
 		return false;
 	}
 
+	U_8 *reserveCheckpointSize() {
```
Inconsistent opening brace placement; e.g. see lines 255, 261, 266, 272, ...

keithc-ca commented on runtime/vm/JFRChunkWriter.hpp:
```diff
@@ -193,6 +194,38 @@ class VM_JFRChunkWriter {
 
+	U_8 *reserveCheckpointSize() {
+		if (J9JAVAVM_COMPRESS_OBJECT_REFERENCES(_vm)) {
+			return _bufferWriter->getAndIncCursor(sizeof(U_32));
+		} else {
+			return _bufferWriter->getAndIncCursor(sizeof(U_64) + sizeof(U_8));
```
The increments should be 4 or 9 (to match the space consumed when the value is actually written).

keithc-ca commented on runtime/vm/JFRChunkWriter.hpp:
```diff
@@ -193,6 +194,38 @@ class VM_JFRChunkWriter {
+		_bufferWriter->writeLEB128PaddedU72(eventStart, (U_64)(_bufferWriter->getCursor() - eventStart));
+	}
+
+	static U_8 *reserveEventSize(VM_BufferWriter *bufferWriter) {
+		return bufferWriter->getAndIncCursor(sizeof(U_64) + sizeof(U_8));
+	}
+
+	static void writeEventSize(VM_BufferWriter *bufferWriter, U_8 *eventStart) {
+		bufferWriter->writeLEB128PaddedU72(eventStart, (U_64)(bufferWriter->getCursor() - eventStart));
+	}
```
Suggest neither of these functions is needed.

thallium replied:
Would you mind explaining why?

keithc-ca replied:
Everywhere this function might be called, it could be replaced by `bufferWriter->writeEventSize(eventStart)`.

keithc-ca replied:
A similar comment applies to `reserveEventSize()`.

tajila replied:
Id like to keep JFR-isms out of `bufferWriter` so we can use in the future for other purposes.

keithc-ca replied:
I agree: My suggestion was in error. Instead, I suggest the non-static methods be implemented via the static methods:
```c
	U_8 *reserveEventSize()
	{
		return reserveEventSize(_bufferWriter);
	}

	void writeEventSize(U_8 *eventStart)
	{
		writeEventSize(_bufferWriter, eventStart);
	}
```

keithc-ca commented on runtime/vm/jfr.cpp:
```diff
@@ -965,16 +965,17 @@ jfrCPULoad(J9VMThread *currentThread)
 				int64_t timeDelta = currentTime - jfrState->prevProcTimestamp;
-				jfrEvent->jvmUser = OMR_MIN((currentProcCPUTimes._userTime - jfrState->prevProcCPUTimes._userTime) / ((double)numberOfCpus * timeDelta), 1.0);
-				jfrEvent->jvmSystem = OMR_MIN((currentProcCPUTimes._systemTime - jfrState->prevProcCPUTimes._systemTime) / ((double)numberOfCpus * timeDelta), 1.0);
+				float timeElapsedAllCores = (float)numberOfCpus * timeDelta;
+				jfrEvent->jvmUser = OMR_MIN((currentProcCPUTimes._userTime - jfrState->prevProcCPUTimes._userTime) / timeElapsedAllCores, 1.0f);
+				jfrEvent->jvmSystem = OMR_MIN((currentProcCPUTimes._systemTime - jfrState->prevProcCPUTimes._systemTime) / timeElapsedAllCores, 1.0f);
```
That's better, but this still (normally) evaluates 
`(currentProcCPUTimes._userTime - jfrState->prevProcCPUTimes._userTime) / timeElapsedAllCores`
twice. Suggest:
```c
				float userTime = (currentProcCPUTimes._userTime - jfrState->prevProcCPUTimes._userTime) / timeElapsedAllCores;
				float systemTime = (currentProcCPUTimes._systemTime - jfrState->prevProcCPUTimes._systemTime) / timeElapsedAllCores;

				jfrEvent->jvmUser = OMR_MIN(userTime, 1.0f);
				jfrEvent->jvmSystem = OMR_MIN(systemTime, 1.0f);
```
A similar issue exists on line 978.

keithc-ca commented on runtime/vm/BufferWriter.hpp:
```diff
@@ -333,6 +337,11 @@ class VM_BufferWriter {
 
+	void writeBoolean(BOOLEAN val)
+	{
+		writeU8((TRUE == val) ? 1 : 0);
```
A boolean should only be tested directly or compared with false:
```c
		writeU8(val ? 1 : 0);
```
(comparing with true is an anti-pattern)

keithc-ca commented on runtime/vm/JFRChunkWriter.cpp:
```diff
@@ -144,8 +144,7 @@ VM_JFRChunkWriter::writeCheckpointEventHeader(CheckpointTypeMask typeMask, U_32
+	_previousCheckpointDelta = reserveEventSize();
 
 	if (_debug) {
 		j9tty_printf(PORTLIB, "next pointer=%d val=%d\n",(int) _bufferWriter->getFileOffsetFromStart(_previousCheckpointDelta), (int) VM_BufferWriter::convertFromLEB128ToU64(_previousCheckpointDelta));
```
Again, casting to `int` is inappropriate.

keithc-ca commented on runtime/vm/JFRChunkWriter.cpp:
```diff
@@ -20,6 +20,7 @@
  * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
  *******************************************************************************/
 #include "JFRUtils.hpp"
+#include "omrformatconsts.h"
```
This is not appropriate. The constants are for use with `printf()`, `sprintf()`, etc, not `omrtty_printf()` (via `j9tty_printf()`).

keithc-ca replied:
It's disappointing to see this still in OMR above the definition of `omrstr_printf()`
```
 * @internal @note Supported, portable format specifiers are described in the document entitled "PortLibrary printf"
 * in the "Inside J9" Lotus Notes database.
```
Here are some sample formats
* `%d` - `int32_t`
* `%lld` - `int64_t`
* `%zd` - `intptr_t`

Have a look at `parseType()` in `omrstr.c` for more details.

thallium replied:
Yes I took a look at the `omrstr.c` and I think I'm using the correct format specifiers now?

keithc-ca commented on runtime/vm/JFRChunkWriter.cpp:
```diff
@@ -77,10 +78,12 @@ U_8 *
 
 	if (_debug) {
-		j9tty_printf(PORTLIB, "Metadata frame start offset = 0x%lX\n", _bufferWriter->getFileOffsetFromStart(dataStart));
+		j9tty_printf(
+			PORTLIB, "Metadata frame start offset = 0x%" OMR_PRIX64 "\n",
+			_bufferWriter->getFileOffsetFromStart(dataStart));
```
When folding, each argument goes on a line by itself (indented two tabs, not one).

keithc-ca commented on runtime/vm/JFRChunkWriter.hpp:
```diff
@@ -24,6 +24,7 @@
 
 #include "j9cfg.h"
 #include "j9.h"
+#include "omrformatconsts.h"
```
Again, this is not appropriate.

thallium replied:
https://github.com/eclipse-openj9/openj9/pull/20158#discussion_r1874049537

keithc-ca replied:
When used with `snprintf()`, yes, but not with `omrstr_printf()` or other OMR API: this should be using OMR functions exclusively for such things.

keithc-ca commented on runtime/vm/jfr.cpp:
```diff
@@ -965,16 +965,20 @@ jfrCPULoad(J9VMThread *currentThread)
 				jfrEvent->machineTotal = 0;
 			} else {
-				jfrEvent->machineTotal = OMR_MIN((currentSysCPUTime.cpuTime - jfrState->prevSysCPUTime.cpuTime) / ((double)numberOfCpus * (currentSysCPUTime.timestamp - jfrState->prevSysCPUTime.timestamp)), 1.0);
+				jfrEvent->machineTotal = OMR_MIN((currentSysCPUTime.cpuTime - jfrState->prevSysCPUTime.cpuTime) / ((float)numberOfCpus * (currentSysCPUTime.timestamp - jfrState->prevSysCPUTime.timestamp)), 1.0f);
```
Please avoid repeated computation due to the use of `OMR_MIN()` like on lines 968-973; also on lines 1011-1012 below.

tajila commented on runtime/vm/JFRChunkWriter.hpp:
```diff
@@ -193,6 +195,44 @@ class VM_JFRChunkWriter {
+	void writeCheckpointSize(U_8 *checkpointEventStart)
+	{
+		if (J9JAVAVM_COMPRESS_OBJECT_REFERENCES(_vm)) {
+			_bufferWriter->writeLEB128PaddedU32(checkpointEventStart, _bufferWriter->getCursor() - checkpointEventStart);
```
Its not the size of the reference that matters, its the size of the checkpoint event. So really the question is, in compressedRefs mode can any checkpoint event be larger than 256Mib? Considering that checkpoint events are threads, methods, classes, classloaders... For most of these it's impossible to consume the full 256Mib. For things like ThreadGroup or methods, its theoretically possible but that is highly unlikely in reality so I think overflow is fine in those cases.

keithc-ca replied:
Then why is it relevant whether the VM uses compressed references or not? Why don't we just use `writeLEB128()` here (unconditionally)?

keithc-ca replied:
256MiB seems small to me. Given the space for the size must be chosen in advance, I suggest we should use `writeLEB128PaddedU72()` (independent of whether compressed references are in use or not).

tajila replied:
> 256MiB seems small to me.

Its large enough for >2million classes. Realistically, we would run out of sub4g memory long before we load 2 million classes. Same goes for threads, classLoaders, packages...

keithc-ca replied:
What kind of "checkpoint" is this related to? I see parts of the old code that use U64, e.g.:
```c
	if (NULL != _previousCheckpointDelta) {
		_bufferWriter->writeLEB128PaddedU64(_previousCheckpointDelta,(U_64) _bufferWriter->getFileOffset(_bufferWriter->getCursor(), _lastDataStart));
```

tajila replied:
Its an offset to the previous checkpoint event

keithc-ca replied:
Ok, but the point is that that allows for 56 bits instead of 28. Why should we restrict offsets as part of fixing "implicit type narrowing"?

I don't think I got a answer to my question: What is a "checkpoint" in this context?

tajila replied:
> Ok, but the point is that that allows for 56 bits instead of 28. Why should we restrict offsets as part of fixing "implicit type narrowing"?

I am not proposing we restrict the size of the offsets. I'm proposing we restrict the size of the checkpoint event `size` field in compressdRefs cases since we dont need more than 28 bits. I see other feedback unrelated to "implicit type narrowing" in this PR.

> I don't think I got a answer to my question: What is a "checkpoint" in this context?

In the `writeCheckpointSize` case. It refers to the size of checkpoint event that is currently being written to.

In the following case: 

```
	if (NULL != _previousCheckpointDelta) {
		_bufferWriter->writeLEB128PaddedU64(_previousCheckpointDelta,(U_64) _bufferWriter->getFileOffset(_bufferWriter->getCursor(), _lastDataStart));

```

It refers to the offset of the start of the previous checkpoint event that was written.

tajila replied:
I spoke we Keith offline. We can use `writeLEB128PaddedU72` everywhere for now, and revisit in the future if it makes sense to use smaller sizes.

keithc-ca commented on runtime/vm/JFRChunkWriter.cpp:
```diff
@@ -144,11 +161,14 @@ VM_JFRChunkWriter::writeCheckpointEventHeader(CheckpointTypeMask typeMask, U_32
-		j9tty_printf(PORTLIB, "next pointer=%d val=%d\n",(int) _bufferWriter->getFileOffsetFromStart(_previousCheckpointDelta), (int) VM_BufferWriter::convertFromLEB128ToU64(_previousCheckpointDelta));
+		j9tty_printf(
+				PORTLIB,
+				"next pointer=%llu val=%llu\n",
```
I don't see an answer to my question about the first parameter format.

tajila replied:
Yes, you can change to hex format

keithc-ca commented on runtime/vm/JFRChunkWriter.hpp:
```diff
@@ -193,6 +194,36 @@ class VM_JFRChunkWriter {
+
+	void writeCheckpointSize(U_8 *checkpointEventStart)
+	{
+		_bufferWriter->writeLEB128PaddedU72(checkpointEventStart, _bufferWriter->getCursor() - checkpointEventStart);
```
Please fold this overly long line:
```c
		_bufferWriter->writeLEB128PaddedU72(
				checkpointEventStart,
				_bufferWriter->getCursor() - checkpointEventStart);
```

keithc-ca commented on runtime/vm/JFRChunkWriter.hpp:
```diff
@@ -193,6 +194,36 @@ class VM_JFRChunkWriter {
+
+	static void writeEventSize(VM_BufferWriter *bufferWriter, U_8 *eventStart)
+	{
+		bufferWriter->writeLEB128PaddedU72(eventStart, (U_64)(bufferWriter->getCursor() - eventStart));
```
Please fold this overly long line and remove the unnecessary cast:
```c
		bufferWriter->writeLEB128PaddedU72(
				eventStart,
				bufferWriter->getCursor() - eventStart);
```

keithc-ca commented on runtime/vm/jfr.cpp:
```diff
@@ -965,16 +965,23 @@ jfrCPULoad(J9VMThread *currentThread)
 				jfrEvent->machineTotal = 0;
 			} else {
-				jfrEvent->machineTotal = OMR_MIN((currentSysCPUTime.cpuTime - jfrState->prevSysCPUTime.cpuTime) / ((double)numberOfCpus * (currentSysCPUTime.timestamp - jfrState->prevSysCPUTime.timestamp)), 1.0);
+				float machineTotal = 
+					(currentSysCPUTime.cpuTime - jfrState->prevSysCPUTime.cpuTime) /
+					((float)numberOfCpus * (currentSysCPUTime.timestamp - jfrState->prevSysCPUTime.timestamp));
```
Lines should be split with a binary operator beginning the continued line:
```c
				float machineTotal =
						(currentSysCPUTime.cpuTime - jfrState->prevSysCPUTime.cpuTime)
						/ ((float)numberOfCpus
						* (currentSysCPUTime.timestamp - jfrState->prevSysCPUTime.timestamp));
```
Continuations should be indented two tabs more than the first line.

keithc-ca commented on runtime/vm/JFRChunkWriter.cpp:
```diff
@@ -144,11 +161,14 @@ VM_JFRChunkWriter::writeCheckpointEventHeader(CheckpointTypeMask typeMask, U_32
-		j9tty_printf(PORTLIB, "next pointer=%d val=%d\n",(int) _bufferWriter->getFileOffsetFromStart(_previousCheckpointDelta), (int) VM_BufferWriter::convertFromLEB128ToU64(_previousCheckpointDelta));
+		j9tty_printf(
+				PORTLIB,
+				"next pointer=%llX val=%llX\n",
```
My comment was only with respect to the first parameter `next pointer=%llX` (which should be `next pointer=0x%llX` (with the leading `0x`) for consistency.

The second parameter was `val=%d`; I was expecting that to become `val=%llu`, but if Tobi prefers hex, it should also have the `0x` prefix.

thallium replied:
Ok I changed the second parameter back to `%llu`.

keithc-ca commented on runtime/vm/JFRChunkWriter.cpp:
```diff
@@ -98,21 +101,32 @@ VM_JFRChunkWriter::writeJFRMetadata()
+		j9tty_printf(
+				PORTLIB,
+				"Metadata blob size from LEB128 = %u\n",
+				VM_BufferWriter::convertFromLEB128ToU32(blobStart));
```
This should use `convertFromLEB128ToU64()` to allow for sizes possibly larger than 28 bits now that writeLEB128PaddedU72()` is used to write size fields.

Same comment for line 129.

keithc-ca commented on runtime/vm/JFRChunkWriter.hpp:
```diff
@@ -240,7 +275,7 @@ class VM_JFRChunkWriter {
 		const size_t fileNameLen = sizeof(intermediateChunkFileName) + 16 + sizeof(".jfr");
 		char fileName[fileNameLen];
-		snprintf(fileName, fileNameLen, "%s%lX.jfr", intermediateChunkFileName, _vm->jfrState.jfrChunkCount);
+		j9str_printf(privatePortLibrary, fileName, fileNameLen, "%s%zX.jfr", intermediateChunkFileName, _vm->jfrState.jfrChunkCount);
```
The `privatePortLibrary` parameter must be removed since #21051.

thallium commented on runtime/vm/JFRChunkWriter.hpp:
```diff
@@ -193,6 +194,40 @@ class VM_JFRChunkWriter {
+
+	static U_8 *reserveEventSize(VM_BufferWriter *bufferWriter)
+	{
+		return bufferWriter->getAndIncCursor(sizeof(U_64) + sizeof(U_8));
```
Yes, at this point I don't think we need them to be distinct. Removed the checkpoint pair.

keithc-ca commented on runtime/vm/JFRChunkWriter.cpp:
```diff
@@ -208,15 +228,15 @@ VM_JFRChunkWriter::writeThreadStateCheckpointEvent()
 
 	/* write size */
-	_bufferWriter->writeLEB128PaddedU32(dataStart, _bufferWriter->getCursor() - dataStart);
+	writeCheckpointSize(dataStart);
```
This has been removed in favour of using `writeEventSize()`.

thallium replied:
Sorry, fixed.

keithc-ca commented:
This still has compile errors on Windows:
```
[ 83%] Building CXX object runtime/vm/CMakeFiles/j9vm.dir/jfr.cpp.obj
openj9\runtime\vm\JFRConstantPoolTypes.hpp(803): warning C4244: 'return': conversion from 'UDATA' to 'U_32', possible loss of data
openj9\runtime\vm\jfr.cpp(1089): warning C4244: '=': conversion from 'UDATA' to 'U_32', possible loss of data
openj9\runtime\vm\jfr.cpp(1090): warning C4244: '=': conversion from 'UDATA' to 'U_32', possible loss of data
openj9\runtime\vm\jfr.cpp(1091): warning C4244: '=': conversion from 'UDATA' to 'U_32', possible loss of data
openj9\runtime\vm\jfr.cpp(1092): warning C4244: '=': conversion from 'UDATA' to 'U_32', possible loss of data
```

thallium commented:
> openj9\runtime\vm\jfr.cpp(1089): warning C4244: '=': conversion from 'UDATA' to 'U_32', possible loss of data
openj9\runtime\vm\jfr.cpp(1090): warning C4244: '=': conversion from 'UDATA' to 'U_32', possible loss of data
openj9\runtime\vm\jfr.cpp(1091): warning C4244: '=': conversion from 'UDATA' to 'U_32', possible loss of data
openj9\runtime\vm\jfr.cpp(1092): warning C4244: '=': conversion from 'UDATA' to 'U_32', possible loss of data

Added explicit casting to these lines.

keithc-ca commented on runtime/vm/jfr.cpp:
```diff
@@ -1076,10 +1086,10 @@ jfrThreadStatistics(J9VMThread *currentThread)
-		jfrEvent->daemonThreadCount = vm->daemonThreadCount;
-		jfrEvent->accumulatedThreadCount = vm->accumulatedThreadCount;
-		jfrEvent->peakThreadCount = vm->peakThreadCount;
+		jfrEvent->activeThreadCount = (U_32)vm->totalThreadCount;
+		jfrEvent->daemonThreadCount = (U_32)vm->daemonThreadCount;
+		jfrEvent->accumulatedThreadCount = (U_32)vm->accumulatedThreadCount;
+		jfrEvent->peakThreadCount = (U_32)vm->peakThreadCount;
```
This is not appropriate: the types of each pair of fields should be the same (or the left side wider). Suggest the JFR types should be updated to match those of the VM, or upgraded to `U_64`. That goes for both `J9JFRThreadStatistics` and `ThreadStatisticsEntry`.

Also, the return type of `VM_JFRConstantPoolTypes::getRequiredBufferSize()` should be `UDATA`.

thallium replied:
At first I thought those fields are meant to be 32 bits since we've been following this [website](https://sap.github.io/SapMachine/jfrevents/21.html#javathreadstatistics) for our fields types but the author for `JavaThreadStatistics` event didn't use the correct type so I'll correct types to `U_64`.

keithc-ca commented on runtime/vm/JFRConstantPoolTypes.hpp:
```diff
@@ -327,27 +327,27 @@ class VM_JFRConstantPoolTypes {
+	U_32 _threadEndCount;
 	J9Pool *_threadSleepTable;
-	UDATA _threadSleepCount;
+	U_32 _threadSleepCount;
 	J9Pool *_monitorWaitTable;
-	UDATA _monitorWaitCount;
+	U_32 _monitorWaitCount;
 	J9Pool *_threadParkTable;
-	UDATA _threadParkCount;
+	U_32 _threadParkCount;
 	J9Pool *_cpuLoadTable;
-	UDATA _cpuLoadCount;
+	U_32 _cpuLoadCount;
 	J9Pool *_threadCPULoadTable;
-	UDATA _threadCPULoadCount;
+	U_32 _threadCPULoadCount;
 	J9Pool *_classLoadingStatisticsTable;
-	UDATA _classLoadingStatisticsCount;
+	U_32 _classLoadingStatisticsCount;
 	J9Pool *_threadContextSwitchRateTable;
 	U_32 _threadContextSwitchRateCount;
 	J9Pool *_threadStatisticsTable;
-	UDATA _threadStatisticsCount;
+	U_32 _threadStatisticsCount;
```
It's not clear to me that we want to change the types of these fields. Please explain the motivation here.

keithc-ca replied:
I think using `UDATA` is more consistent with data-type usage elsewhere in the VM, but, of course the getters need to be consistent, so they must return `UDATA` as well.

keithc-ca commented on runtime/vm/JFRConstantPoolTypes.hpp:
```diff
@@ -19,6 +19,7 @@
  *
  * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
  *******************************************************************************/
+#include "omrcomp.h"
```
Please don't add this. It's already included via `omr.h`.

keithc-ca commented on runtime/vm/JFRConstantPoolTypes.hpp:
```diff
@@ -19,6 +19,7 @@
  * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
  *******************************************************************************/
+#include "omrcomp.h"
 #include <cstring>
```
This belongs in `.cpp` files where needed (if at all).

keithc-ca replied:
Please move the include of `<cstring>` after the includes of openj9 and omr header files (it should be within the `!defined(JFRCONSTANTPOOLTYPES_HPP_)` section).

keithc-ca commented on runtime/vm/JFRConstantPoolTypes.hpp:
While you're here, please correct the comment on the last line:
```c
#endif /* !defined(JFRCONSTANTPOOLTYPES_HPP_) */
```

keithc-ca commented on runtime/vm/JFRConstantPoolTypes.hpp:
```diff
@@ -19,10 +19,11 @@
 #if !defined(JFRCONSTANTPOOLTYPES_HPP_)
 #define JFRCONSTANTPOOLTYPES_HPP_
 
+#include <cstring>
```
Please move this *after* the openj9 and omr includes, as requested; then squash.

keithc-ca commented on runtime/vm/JFRConstantPoolTypes.hpp:
```diff
@@ -32,6 +31,8 @@
 #include "vm_internal.h"
 #include "ut_j9vm.h"
 
+#include <cstring>
+
 #if defined(J9VM_OPT_JFR)
```
Please move the include of `cstring` after all openj9 and omr includes (after `VMHelpers.hpp`).
Also, line 36 should be moved before the include of `j9hypervisor.h`.

```c
#include "j9cfg.h"

#if defined(J9VM_OPT_JFR)

#include "j9hypervisor.h"
#include "j9.h"
#include "omr.h"
#include "omrlinkedlist.h"
#include "vm_api.h"
#include "vm_internal.h"
#include "ut_j9vm.h"
#include "BufferWriter.hpp"
#include "JFRUtils.hpp"
#include "ObjectAccessBarrierAPI.hpp"
#include "VMHelpers.hpp"

#include <cstring>
```

keithc-ca commented:
It doesn't appear there will be any Windows build machines online anytime soon, so I tried a local build. These errors will need to be addressed:
```
openj9\runtime\vm\JFRConstantPoolTypes.cpp(965): warning C4244: '=': conversion from 'UDATA' to 'U_32', possible loss of data
openj9\runtime\vm\JFRConstantPoolTypes.cpp(997): warning C4244: '=': conversion from 'UDATA' to 'U_32', possible loss of data
openj9\runtime\vm\JFRConstantPoolTypes.cpp(1023): warning C4244: '=': conversion from 'UDATA' to 'U_32', possible loss of data
openj9\runtime\vm\JFRConstantPoolTypes.cpp(1054): warning C4244: '=': conversion from 'UDATA' to 'U_32', possible loss of data
openj9\runtime\vm\JFRConstantPoolTypes.cpp(1092): warning C4244: '=': conversion from 'UDATA' to 'U_32', possible loss of data
openj9\runtime\vm\JFRConstantPoolTypes.cpp(1151): warning C4244: '=': conversion from 'UDATA' to 'U_32', possible loss of data
openj9\runtime\vm\JFRConstantPoolTypes.cpp(1178): warning C4244: '=': conversion from 'UDATA' to 'U_32', possible loss of data
openj9\runtime\vm\JFRConstantPoolTypes.cpp(1200): warning C4244: '=': conversion from 'UDATA' to 'U_32', possible loss of data
```
Please ensure such builds succeed before you bring this out of the draft state.

thallium commented:
@keithc-ca build passed on Windows https://hyc-runtimes-jenkins.swg-devops.com/job/Pipeline-Build-Test-Personal/26418/

thallium commented:
Changed the return type of `VM_JFRConstantPoolTypes::add<jfr event>Entry()` to void as the returned index is no longer needed. Removed the `index` field of `ExecutionSampleEntry` as it's not used anywhere.

keithc-ca commented on runtime/vm/JFRConstantPoolTypes.cpp:
```diff
@@ -509,7 +509,7 @@ VM_JFRConstantPoolTypes::addPackageEntry(J9Class *clazz)
 	}
 
 	entry->packageName = (U_8*) packageName;
```
The type of the field `packageName` and the local of the same name should be `const U_8 *` so this cast and the one on line 505 would not be needed. The cast on line 1280 is missing the `const` qualifier and that line should be split:
```c
		j9tty_printf(
				PORTLIB,
				"%u) moduleIndex=%u packageName=%.*s exported=%u\n",
				tableEntry->index,
				tableEntry->moduleIndex,
				tableEntry->packageNameLength,
				(const char *)tableEntry->packageName,
				tableEntry->exported);
```

keithc-ca commented on runtime/vm/JFRChunkWriter.hpp:
The `const` qualifier is missing in the signature of the `writeUTF8String()` methods:
```c
	void writeUTF8String(const J9UTF8 *string);

	void writeUTF8String(const U_8 *data, UDATA len);
```

keithc-ca commented on runtime/vm/JFRChunkWriter.hpp:
```diff
@@ -682,9 +705,9 @@ class VM_JFRChunkWriter {
 	U_8 *writeCheckpointEventHeader(CheckpointTypeMask typeMask, U_32 cpCount);
 
-	void writeUTF8String(J9UTF8* string);
+	void writeUTF8String(const J9UTF8* string);
```
Space before, not after, `*`.

keithc-ca commented on runtime/vm/JFRConstantPoolTypes.hpp:
```diff
@@ -93,7 +94,7 @@ struct PackageEntry {
 	BOOLEAN exported;
 	U_32 packageNameLength;
-	U_8* packageName;
+	const U_8* packageName;
```
Space before, not after, `*`.

keithc-ca commented:
Jenkins test sanity.functional alinux,osx jdk21
